install/kubernetes/helm/istio/charts/gateways/templates/preconfigured.yaml

```diff
@@ -0,0 +1,155 @@
+{{- if .Values.global.k8sIngress.enabled }}
+apiVersion: networking.istio.io/v1alpha3
+kind: Gateway
+metadata:
+  name: istio-autogenerated-k8s-ingress
+  namespace: {{ .Release.Namespace }}
+spec:
+  selector:
+    istio: {{ .Values.global.k8sIngress.gatewayName }}
+  servers:
+  - port:
+      number: 80
+      protocol: HTTP2
+      name: http
+    hosts:
+    - "*"
+{{ if .Values.global.k8sIngress.enableHttps }}
+  - port:
+      number: 443
+      protocol: HTTPS
+      name: https-default
+    tls:
+      mode: SIMPLE
+      serverCertificate: /etc/istio/ingress-certs/tls.crt
+      privateKey: /etc/istio/ingress-certs/tls.key
+    hosts:
+    - "*"
+{{ end }}    
+---    
+{{ end }}
+
+{{- if .Values.global.meshExpansion.enabled }}
+{{- if .Values.global.meshExpansion.useILB }}
+apiVersion: networking.istio.io/v1alpha3
+kind: Gateway
+metadata:
+  name: meshexpansion-ilb-gateway
+  namespace: {{ .Release.Namespace }}
+spec:
+  selector:
+    istio: ilbgateway
+  servers:
+  - port:
+      number: 15011
+      protocol: TCP
+      name: tcp-pilot
+    hosts:
+    - "*"
+  - port:
+      number: 8060
+      protocol: TCP
+      name: tcp-citadel
+    hosts:
+    - "*"
+---
+{{- else }}
+apiVersion: networking.istio.io/v1alpha3
+kind: Gateway
+metadata:
+  name: meshexpansion-gateway
+  namespace: {{ .Release.Namespace }}
+spec:
+  selector:
+    istio: ingressgateway
+  servers:
+  - port:
+      number: 15011
+      protocol: TCP
+      name: tcp-pilot
+    hosts:
+    - "*"
+  - port:
+      number: 8060
+      protocol: TCP
+      name: tcp-citadel
+    hosts:
+    - "*"
+---
+{{- end }}
+{{- end }}
+
+{{- if .Values.global.multiCluster.connectUsingGateway }}
+apiVersion: networking.istio.io/v1alpha3
+kind: Gateway
+metadata:
+  name: istio-multicluster-gateway
+  namespace: {{ .Release.Namespace }}
+spec:
+  selector:
+    istio: {{ .Values.global.multiCluster.gatewayName }}
+  servers:
+  - hosts:
+    # This is a dummy port to keep AWS LB happy.
+    # DO NOT REMOVE.
+    - "*.local"
+    port:
+      name: http
+      number: 80
+      protocol: HTTP
+  - hosts:
+    - "*.local"
+    port:
+      name: tls
+      number: 443
+      protocol: TLS
+    tls: {}
+---
+apiVersion: networking.istio.io/v1alpha3
+kind: VirtualService
+metadata:
+  name: istio-multicluster-virtualservice
+  namespace: {{ .Release.Namespace }}
+spec:
+  gateways:
+  - istio-multicluster-gateway
+  hosts:
+  - "*.local"
+  http: #black hole all port 80 traffic.
+  - match:
+    - port: 80
+    route:
+    - destination:
+        host: non.existent.cluster
+      weight: 100
+  # We need a TLS block to setup the filter chain
+  # The actual routing happens using the SNI filter
+  tls:
+  - match:
+    - port: 15443
+      sniHosts:
+      - "*.local"
+    route:
+    - destination:        
+        host: non.existent.cluster
```

Comment:
I was trying to use this config. The code to build gateway listeners complains about non existent service (which is non.existent.cluster). And because of that, listener for port 15443 was not added into the gw config. I changed the destination host to one that exists in the cluster, and the listener was added. But I guess that's not needed if this PR works.

Secondly, the match port in here is 15443, while in the gateway spec, it is 443 (at line 104). Is it correct? I thought that they should be the same port number.

In general, we need to have a better understanding on how sni_cluster works. The envoy spec W.R.T sni_cluster is very brief, which seemed to be added very recently.